cmd/cli/main.go

```diff
@@ -65,6 +67,89 @@ var (
 	pipelineId           string
 )
 
+func getRemoteHash(repoUrl string, branch string) (string, error) {
+	cmd := exec.Command("git", "ls-remote", repoUrl, branch)
+	var out bytes.Buffer
+	var stderr bytes.Buffer
+	cmd.Stdout = &out
+	cmd.Stderr = &stderr
+	err := cmd.Run()
+	if err != nil {
+		return "", fmt.Errorf("%s: %s", err, stderr.String())
+	}
+	parts := strings.Fields(out.String())
+	if len(parts) > 0 {
+		return parts[0], nil
+	}
+	return "", fmt.Errorf("repo or branch not found")
+}
+
+func copyDir(src string, dst string) error {
+	cmd := exec.Command("cp", "-r", src, dst)
+	return cmd.Run()
+}
+
+func syncRepo(repoUrl string, branch string, targetDir string) error {
+	hasher := sha1.New()
+	hasher.Write([]byte(repoUrl))
+	repoHash := hex.EncodeToString(hasher.Sum(nil))
+	cacheDir := homeDir + "/.irgsh/cache/" + repoHash
+
+	remoteHash, err := getRemoteHash(repoUrl, branch)
+	if err != nil {
+		return err
+	}
+
+	if _, err := os.Stat(cacheDir); !os.IsNotExist(err) {
+		repo, err := git.PlainOpen(cacheDir)
+		if err == nil {
+			ref, err := repo.Head()
+			if err == nil {
+				if ref.Hash().String() == remoteHash {
+					log.Println("[syncRepo] cache hit for " + repoUrl)
+					return copyDir(cacheDir, targetDir)
+				}
+			}
+		}
+		log.Println("[syncRepo] cache stale, updating...")
+		worktree, err := repo.Worktree()
+		if err == nil {
+			err = worktree.Pull(&git.PullOptions{
+				RemoteName:    "origin",
+				ReferenceName: plumbing.ReferenceName(fmt.Sprintf("refs/heads/%s", branch)),
+				SingleBranch:  true,
+			})
+			if err != nil && err != git.NoErrAlreadyUpToDate {
+				os.RemoveAll(cacheDir)
+			} else {
+				return copyDir(cacheDir, targetDir)
```

Comment:
Please try to avoid 
1. Direct return like above
2. Using `err == nil` unless the case is very specific.

Verbosity is power in golang.

```
  err = copyDir()
  if err != nil {
    // handle the error (log and early return)
  }

  // Then return the result
```
This way runtime will tell you which line that causes the error.